src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php

```diff
@@ -429,72 +366,64 @@ private function createNewInvoice(Order $order, Cart $cart, $isFreeShipping, arr
-        $orderCarrier->add();
+        if (true === $withFreeShipping) {
+            $this->addFreeShippingDiscount($order, $cart, $invoice);
+        }
 
         return $invoice;
     }
 
     /**
-     * @param int $orderInvoiceId
+     * @param Order $order
      * @param Cart $cart
-     * @param array $newProducts
+     * @param OrderInvoice $orderInvoice
      *
-     * @return OrderInvoice
+     * @throws OrderException
      */
-    private function updateExistingInvoice($orderInvoiceId, Cart $cart, array $newProducts)
+    private function addFreeShippingDiscount(Order $order, Cart $cart, OrderInvoice $orderInvoice): void
     {
-        $invoice = new OrderInvoice($orderInvoiceId);
+        $defaultLangId = (int) Configuration::get('PS_LANG_DEFAULT');
+        $cartRuleObj = new CartRule();
+        $cartRuleObj->id_customer = $order->id_customer;
+        $cartRuleObj->date_from = date('Y-m-d H:i:s', strtotime('-1 hour', strtotime($order->date_add)));
+        $cartRuleObj->date_to = date('Y-m-d H:i:s', strtotime('+1 hour'));
+        $cartRuleObj->name = [
+            $defaultLangId => $this->translator->trans(
+                'Free Shipping discount for invoice %s',
+                ['%s' => $orderInvoice->getInvoiceNumberFormatted($defaultLangId, $order->id_shop)],
+                'Admin.Orderscustomers.Notification'
+            ),
+        ];
+        // This a one time cart rule, for a specific user that can only be used once
+        $cartRuleObj->quantity = 1;
+        $cartRuleObj->quantity_per_user = 1;
+        $cartRuleObj->active = 0;
+        $cartRuleObj->highlight = 0;
+        $cartRuleObj->id_order_invoice = $orderInvoice->id;
+        $cartRuleObj->free_shipping = 1;
 
-        $invoice->total_paid_tax_excl += Tools::ps_round(
-            (float) $cart->getOrderTotal(false, Cart::BOTH_WITHOUT_SHIPPING, $newProducts),
-            $this->computingPrecision
-        );
-        $invoice->total_paid_tax_incl += Tools::ps_round(
-            (float) $cart->getOrderTotal(true, Cart::BOTH_WITHOUT_SHIPPING, $newProducts),
-            $this->computingPrecision
-        );
-        $invoice->total_products += (float) $cart->getOrderTotal(
-            false,
-            Cart::ONLY_PRODUCTS,
-            $newProducts
-        );
-        $invoice->total_products_wt += (float) $cart->getOrderTotal(
-            true,
-            Cart::ONLY_PRODUCTS,
-            $newProducts
-        );
-
-        $invoice->update();
+        try {
+            if (!$cartRuleObj->add()) {
+                throw new OrderException('An error occurred during the CartRule creation');
+            }
+        } catch (PrestaShopException $e) {
+            throw new OrderException('An error occurred during the CartRule creation', 0, $e);
+        }
 
-        return $invoice;
+        try {
+            // It's important to add the cart rule to the cart Or it will be ignored when cart performs AutoRemove AddAdd
```

Comment:
`AutoRemove AddAdd` ? 🤔

Reply:
Yes this a legacy mechanism that is internal to the Cart, by default when you call `$cart->getCartRules()` it performs an automatic add of CartRules that may be associated to it, as long as they meet the conditions

This is this mechanism that will automatically add a discount or a free shipping when you reach a minimum amount in your cart for example. Similarly there is an autoRemove feature which is called when you remove a product from the cart or update its quantity.

I didn't know of these two mechanisms before either, but they are deeply connected to the cart rules behaviour, and can not be easily disabled. The autoAdd can be ignored if you force to false when calling `getCartRules` but it doesn't prevent any other module or hook to call the same method which will automatically add them later. And the `autoRemove` is simply just not possible to be disabled.

So instead of fighting against these two mechanisms it's better to go with them, this is what has been done in the recent fixes to Order cart rules, we don't compute them manually (which resulted in many computations errors), instead we rely on the cart intelligence to add the good cart rules, compute them correctly as it would do in the front and only then we synchronize OrderCartRules based on the computed cart rules.

So we "just" need to correctly initialize the CartRules and add them to the cart to profit from this mechanism.